gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -104,34 +108,50 @@
    */
   @VisibleForTesting
   static final String KAFKA_TOPIC_SPECIFIC_STATE = "kafka.topic.specific.state";
+  public static final String KAFKA_SOURCE_WORK_UNITS_CREATION_THREADS = "kafka.source.work.units.creation.threads"; 
+  public static final int DEFAULT_THREAD_COUNT = 10;
 
   private final Set<String> moveToLatestTopics = Sets.newTreeSet(String.CASE_INSENSITIVE_ORDER);
-  private final Map<KafkaPartition, Long> previousOffsets = Maps.newHashMap();
+  private final Map<KafkaPartition, Long> previousOffsets = Maps.newConcurrentMap();
 
-  private final Set<KafkaPartition> partitionsToBeProcessed = Sets.newHashSet();
+  private final Set<KafkaPartition> partitionsToBeProcessed = Sets.newConcurrentHashSet();
 
   private Closer closer = Closer.create();
   private KafkaWrapper kafkaWrapper;
-  private int failToGetOffsetCount = 0;
-  private int offsetTooEarlyCount = 0;
-  private int offsetTooLateCount = 0;
+  private final AtomicInteger failToGetOffsetCount = new AtomicInteger(0);
+  private final AtomicInteger offsetTooEarlyCount = new AtomicInteger(0);
+  private final AtomicInteger offsetTooLateCount = new AtomicInteger(0);
+  private final AtomicBoolean doneGettingAllPreviousOffsets = new AtomicBoolean(false);
 
   @Override
   public List<WorkUnit> getWorkunits(SourceState state) {
-    Map<String, List<WorkUnit>> workUnits = Maps.newHashMap();
+    Map<String, List<WorkUnit>> workUnits = Maps.newConcurrentMap(); 
 
     this.kafkaWrapper = this.closer.register(KafkaWrapper.create(state));
 
     List<KafkaTopic> topics = getFilteredTopics(state);
     Map<String, State> topicSpecificStateMap = getTopicSpecificState(topics, state);
+    
+    int numOfThreads =
+        state.getPropAsInt(KAFKA_SOURCE_WORK_UNITS_CREATION_THREADS, DEFAULT_THREAD_COUNT);
+    ExecutorService threadPool = Executors.newFixedThreadPool(numOfThreads,
+        ExecutorsUtils.newThreadFactory(Optional.of(LOG)));
+    
+    if(topics != null){
+      LOG.info("Begin using thread pool to create work units for topic size " + topics.size());
+    }
+    
     for (KafkaTopic topic : topics) {
-      workUnits.put(topic.getName(),
-          getWorkUnitsForTopic(topic, state, Optional.fromNullable(topicSpecificStateMap.get(topic.getName()))));
+      LOG.info("Using thread pool to create work units for topic " + topic.getName());
+      threadPool.submit(new WorkUnitCreator(topic, state, Optional.fromNullable(topicSpecificStateMap.get(topic.getName())), this, workUnits));
     }
+    
```

Comment:
Maybe useful to log a message here. We will know how long we are waiting for all the threads to complete.